internal/db/p2p/sync_branchable_col.go

```diff
@@ -0,0 +1,299 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package p2p
+
+import (
+	"context"
+	"time"
+
+	"github.com/fxamacker/cbor/v2"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime/linking"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+
+	"github.com/sourcenetwork/corelog"
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/errors"
+	"github.com/sourcenetwork/defradb/event"
+	coreblock "github.com/sourcenetwork/defradb/internal/core/block"
+	"github.com/sourcenetwork/defradb/internal/datastore"
+	dbid "github.com/sourcenetwork/defradb/internal/db/id"
+	"github.com/sourcenetwork/defradb/internal/keys"
+)
+
+// syncBranchableCollectionTopic is the fixed topic for branchable collection sync operations.
+const syncBranchableCollectionTopic = "sync-branchable"
+
+// syncBranchableCollectionRequest represents a request to synchronize a branchable collection.
+type syncBranchableCollectionRequest struct {
+	CollectionID string `json:"collectionID"`
+}
+
+// syncBranchableCollectionReply represents the response to a collection sync request.
+type syncBranchableCollectionReply struct {
+	CollectionID string   `json:"collectionID"`
+	Heads        [][]byte `json:"heads"`
+	Sender       string   `json:"sender"`
+}
+
+// SyncBranchableCollection initiates a request for the latest version of the branchable
+// collection's DAG from the network.
+//
+// This function call will block until there is a response for the collection.
+// It is the responsibility of the caller to set an appropriate timeout on the context.
+func (p *P2P) SyncBranchableCollection(ctx context.Context, collectionID string) error {
+	cols, err := p.db.GetCollections(
+		ctx,
+		client.CollectionFetchOptions{
+			CollectionID: immutable.Some(collectionID),
+		},
+	)
+	if err != nil {
+		return err
+	}
+	if len(cols) == 0 {
+		return client.NewErrCollectionNotFoundForCollectionVersion(collectionID)
+	}
+
+	col := cols[0].Version()
+	if !col.IsBranchable {
+		return NewErrCollectionNotBranchable(collectionID)
+	}
+
+	return p.syncBranchableCollection(ctx, collectionID)
+}
+
+// syncBranchableCollection requests branchable collection synchronization from the network.
+func (p *P2P) syncBranchableCollection(
+	ctx context.Context,
+	collectionID string,
+) error {
+	pubsubReq := &syncBranchableCollectionRequest{CollectionID: collectionID}
+
+	data, err := cbor.Marshal(pubsubReq)
+	if err != nil {
+		return err
+	}
+
+	pubSubRespChan, err := p.host.PublishToTopic(ctx, syncBranchableCollectionTopic, data, true)
+	if err != nil {
+		return err
+	}
+
+	waitCtx := ctx
+	if _, hasDeadline := ctx.Deadline(); !hasDeadline {
+		var cancel context.CancelFunc
+		waitCtx, cancel = context.WithTimeout(ctx, 5*time.Second)
+		defer cancel()
+	}
+
+	return p.waitAndHandleSyncBranchableCollectionResponse(waitCtx, collectionID, pubSubRespChan)
```

Comment:
question: This function looks like it will always block for 5 seconds, unless the user has already provided their own timeout, or an error is returned.  Once the context timeout has been breached, syncing will cease, even if more responses have been received locally and are queued for processing (the processing appears limited to a single routine, so they will be handled one at a time).

This feels inconsistent with the user-visible documentation, which simply notes that a timeout may be provided.  This is not how the collection version sync works.  How does doc sync work?

Have I misread the code?  Is this the behaviour we want? 

todo: If it does work like this, and we want this, I think the user-visible documentation needs to be clearer.